### ISSUE TITLE
Small Problem

### ISSUE THREAD

**Issue by nightbreezecool** [kkakkkka/ETRIS#17]
**Can you provide the code to calculate validate loss ?**

**Comment by kkakkkka** [comment]
Hello,
﻿
Thank you for your patience and understanding regarding our project. Our project actually doesn't use this validate loss, but I can help you implement it later. I expect to have some time in late January 2025.
﻿
Thank you for your continued support and interest in our project.

**Comment by kkakkkka** [comment]
To calculate the validation loss, you need to modify the `RefDataset` to return the ground truth masks during validation, and then update the `validate` function in `engine.py` to compute the loss.

Here are the necessary changes:

### 1. Modify `utils/dataset.py`

In the `__getitem__` method, update the `elif self.mode == 'val':` block. We need to load and transform the mask just like we do in training mode so it can be batched and used for loss calculation.

**File:** `utils/dataset.py`

```python
# ... (inside __getitem__)
        elif self.mode == 'val':
            # sentence -> vector
            sent = sents[0]
            word_vec = tokenize(sent, self.word_length, True).squeeze(0)
            
            # === NEW CODE START: Load and Transform Mask ===
            # Load mask from LMDB
            mask = cv2.imdecode(np.frombuffer(ref['mask'], np.uint8),
                                cv2.IMREAD_GRAYSCALE)
            # Apply the same affine transform as the image
            mask = cv2.warpAffine(mask,
                                  mat,
                                  self.input_size,
                                  flags=cv2.INTER_LINEAR,
                                  borderValue=0.)
            mask = mask / 255.
            
            # Convert both image and mask to Tensors
            img, mask = self.convert(img, mask)
            # === NEW CODE END ===

            # (Remove or comment out the old line: img = self.convert(img)[0])
            
            params = {
                'mask_dir': mask_dir,
                'inverse': mat_inv,
                'ori_size': np.array(img_size)
            }
            # Return mask as well
            return img, word_vec, mask, params 

```

### 2. Modify `engine/engine.py`

Update the `validate` function to accept the `mask` from the data loader, calculate the Binary Cross Entropy (BCE) loss, and log it.

**File:** `engine/engine.py`

```python
@torch.no_grad()
def validate(val_loader, model, epoch, args):
    iou_list = []
    
    # === NEW: Initialize Loss Meter ===
    loss_meter = AverageMeter('Loss', ':2.4f')
    
    model.eval()
    time.sleep(2)
    
    # === UPDATE: Unpack masks from loader ===
    for imgs, texts, masks, param in val_loader:
        # data
        imgs = imgs.cuda(non_blocking=True)
        texts = texts.cuda(non_blocking=True)
        # Add channel dimension to mask: (B, H, W) -> (B, 1, H, W)
        masks = masks.cuda(non_blocking=True).unsqueeze(1) 

        # inference
        preds = model(imgs, texts)
        
        # === NEW: Calculate Validation Loss ===
        # Resize mask to match prediction size if necessary (handling strides)
        if preds.shape[-2:] != masks.shape[-2:]:
            target_mask = F.interpolate(masks, preds.shape[-2:], mode='nearest')
        else:
            target_mask = masks
            
        # Calculate BCE Loss (preds are logits)
        loss = F.binary_cross_entropy_with_logits(preds, target_mask)

        # Aggregate loss across GPUs (if distributed)
        dist.all_reduce(loss.detach())
        loss = loss / dist.get_world_size()
        loss_meter.update(loss.item(), imgs.size(0))
        # ======================================

        preds = torch.sigmoid(preds)
        # ... (rest of the existing IoU calculation code) ...
        # (Note: Keep the rest of the loop logic for IoU calculation unchanged)

    # ... (existing aggregation code) ...

    # === UPDATE: Log the Loss ===
    head = 'Evaluation: Epoch=[{}/{}]  Loss={:.4f} IoU={:.2f}'.format(
        epoch, args.epochs, loss_meter.avg, 100. * iou.item())
    logger.info(head + temp)
    
    return iou.item(), prec

```

With these changes, the validation loop will now compute and print the `Loss` alongside the IoU metrics.
